### PR TITLE
bigslice.Repartition: partition slice with a user function 

### PR DESCRIPTION
**Pull request by mariusae**
This change introduces the combinator bigslice.Repartition.
bigslice.Repartition is like bigslice.Refshuffle, except that
it takes a user-provided partitioning function, allowing
fine-grained user control of sharding.

This is the rest of the implementation required for #17.

### CODE REVIEW

File: reshuffle.go [hunk 1]
```diff
@@ -28,12 +38,47 @@ func Reshuffle(slice Slice) Slice {
...
+	}
+	fval := slicefunc.Of(fn)
+	part := func(frame frame.Frame, nshard int, shards []int) {
+		args := make([]reflect.Value, slice.NumOut()+1)
+		args[0] = reflect.ValueOf(nshard)
+		for i := range shards {
+			for j := 0; j < slice.NumOut(); j++ {
+				args[j+1] = frame.Index(j, i)
+			}
+			result := fval.Call(context.TODO(), args)
```
Comment (cosnicolaou) [inline reshuffle.go]: for the context.TODO(), is it your intention to use a context as for counters here too? That is, if a context is provided, then use it in the Call also. If so, probably a good idea to say so and to add an issue for cleaning these up in the future.

Comment (mariusae) [reply reshuffle.go]: yes, that's the plan: it requires some internal changes to plumb through contexts. i'll add a note.

Comment (mariusae) [reply reshuffle.go]: actually it was just quicker to fix it right away.

File: reshuffle.go [hunk 2]
```diff
@@ -28,12 +38,47 @@ func Reshuffle(slice Slice) Slice {
 	if err := canMakeCombiningFrame(slice); err != nil {
 		typecheck.Panic(1, err.Error())
 	}
-	return &reshuffleSlice{makeName("reshuffle"), slice}
+	return &reshuffleSlice{makeName("reshuffle"), nil, slice}
+}
+
+// Repartition (re-)partitions the slice according to the provided function
+// fn, which is invoked for each record in the slice to assign that record's
+// shard. The function is is supplied with the number of shards to partition
```
Comment (jcharum) [inline reshuffle.go]:
```suggestion
// shard. The function is supplied with the number of shards to partition
```

**Pull request merged**